src/language/patterns.md

```diff
@@ -0,0 +1,334 @@
+and extract it in parts. In other words, the pattern _destructures_ the object:
+
+```dart
+var numList = [1, 2, 3];
+var [a, b, c] = numList;   // List pattern [a, b, c] destructures the list object
+print(a + b + c);          // numList into individual new parts. Prints 6.
+```
+
+You can use the [null-check pattern][] to check for `null` while destructuring. 
+
+All patterns attempt to match the objects they're compared against, but not all
+patterns destructure.
+Whether a pattern destructures depends on the [type of pattern][types]
+and the [context](#pattern-uses) in which it's being used.
+
+## Pattern uses
+
+Patterns integrate into several code constructs across the Dart language:
+
+- Local variable [declarations](#variable-declaration) and [assignments](#variable-assignment)
+- [for and for-in loops][for]
+- [if-cases] and [switch-cases]
+- [Collection literals][]
+
+This section describes common use cases for matching and destructuring with patterns, 
+combining various contexts, pattern types, and surrounding code constructs.
+
+### Variable declaration
+
+Anywhere Dart allows regular, single-name local variable declaration, you can also
+use a _pattern variable declaration_. If the pattern matches the value on the right
+of the declaration, it destructures the value, binding it to new local variables:
+
+```dart
+var (a, [b, c]) = ("str", [1, 2]);   // Declares new variables a, b, and c.
+```
+
+A pattern variable declaration must start with either `var` or `final`, followed
+by a pattern. 
+
+Use the [null-assert pattern][] to eliminate `null` in variable declarations.
+
+### Variable assignment 
+
+A _variable assignment pattern_ on the left side of an assignment destructures the
+matched object and assigns the values to _existing_ variables,
+instead of binding new ones. 
+
+One convenient use case is swapping the values of two variables without needing
+to declare a third temporary one:
+
+```dart
+var (a, b) = ('left', 'right');
+(b, a) = (a, b);                   // Swap
+print('$a $b');                    // Prints "right left".
+```
+
+### Destrcuturing multiple returns
+
+[Records][] allow aggregating and returning multiple values from a single function
+call. Patterns add the ability to destructure a record's fields
+directly into local variables, inline with the function call.
+
+Instead of individually declaring new local variables for each record field,
+like this:
+
+```dart
+main() {
+  var info = userInfo(json);
+  var name = info.$1;
+  var age = info.$2;
+}
+```
+You can call the function directly into a [variable declaration](#variable-declaration)
+or [assigment pattern](#variable-assignment),
+with a [record subpattern][record] and fields for each value in the return:
+
+```dart
+main() {
+  var (name, age) = userInfo(json);
+}
+```
+
+### Switch cases
+
+The `case` keyword is always followed by a pattern. _Case patterns_ are [refutable][];
```

Comment:
This is true but doesn't also cover switch *expressions* where the pattern isn't preceded by `case`. (It was in early versions of the proposal, but we changed the syntax.)

Maybe say "Every case in a switch statement or expression contains a pattern."